Do not render password value

Martin-P opened the pull request:
This PR fixes #7166. The original issue #2602 was partially solved in #2613. I used the same approach as in #2613.

Ocramius commented on library/Zend/Form/Element.php:
```diff
@@ -83,6 +86,20 @@ public function init()
+    public function prepareElement(FormInterface $form)
+    {
+        if ('password' == $this->getAttribute('type')) {
+            $this->setValue('');
```
This seems wrong to me: you don't know when `prepareElement` is going to be called. Instead, it would be better to filter out the value before rendering

Martin-P replied:
It is always called at `$form->prepare()` and because `$form->prepare()` is needed before rendering a form this will work just fine.

I agree ideally you should alter the view helper, but because an `<input type="password" />` can be an instance of `Zend\Form\Element` *or* `Zend\Form\Element\Password` and thus can be rendered by `Zend\Form\View\Helper\FormInput` *or* `Zend\Form\View\Helper\FormPassword` it needs fixing on two separate places. I am opening a new issue about this, because it affects all form elements.

Martin-P replied:
Please see #7172 for a more thorough explanation.

Ocramius replied:
@Martin-P I know when it is called, but that's still a big mess if anything modifies the form after `prepare()`. We rely on that initializer approach too much, and that makes the component highly unstable.

Martin-P replied:
What would be your suggestion? Changing the view helpers? Can do that, no problem, but placing password specific code in the generic `FormInput` view helper does not seem the right way to do this...

[Line 124 `Zend\Form\View\Helper\FormInput`](https://github.com/zendframework/zf2/blob/67f098af070b29d5042e89e936604df3193d2212/library/Zend/Form/View/Helper/FormInput.php#L124)
Change this:
```php
$attributes['type']  = $this->getType($element);
$attributes['value'] = $element->getValue();
```
Into this?
```php
$type                = $this->getType($element);
$attributes['type']  = $type;
$attributes['value'] = $element->getValue();
if ('password' == $type) {
    $attributes['value'] = '';
}
```

Martin-P replied:
Ping @Ocramius Im adding a tag `awaiting organisation updates` :wink:

weierophinney replied:
@Martin-P Actually, yes, I think that _would_ make more sense. There are cases where you want to keep the value of the password in the element, even if you do not want to _display_ it. As such, doing this in the view layer makes more sense, as display issues are view issues.

Martin-P committed:
Password value should never be rendered

Martin-P commented:
@weierophinney I updated the view helper so it does not render the password value value anymore and also added tests for both angles:
-  `Zend\Form\View\Helper\FormInput`
- `Zend\Form\View\Helper\FormPassword`

This PR makes PR #2613 obsolete, so theoretically that patch can be removed.

weierophinney commented:
Merged to develop for release with 2.4.
